### PR TITLE
Add "useLocalMedia" hook for local media handling

### PR DESCRIPTION
**Pull request by havardholvik**
Expose a new hook to allow management of local media (audio, video).

### Tested like
1. `STORYBOOK_ROOM=<unlocked_room_in_production> yarn storybook`
2. Join your unlocked room as normal in one tab
3. Go to http://localhost:6006/?path=/story/examples-custom-ui--local-media-only
   - Switch between local cameras, verify the preview updates
   - Toggle camera on off, verify you get black screen
   - Navigate to the "Start / Stop" story to verify your camera lights are turned off
4. Go to http://localhost:6006/?path=/story/examples-custom-ui--room-connection-with-local-media
   - Choose camera and microphone, join room
   - Verify you get audio and video
   - Try using either toggle button to verify you are able to toggle cam / mic, and state is updated correctly for the remote participant
   - Try changing devices, verify the new device is being used in the call
   - Leave room by going to the "Start / Stop" story again, verify camera lights turn off

Todo:
- [x] Handle GUM error
- [x] Support toggling camera + microphone
- [x] Support enumerating devices
- [x] Support switching devices

### CODE REVIEW

File: src/lib/RoomConnection.ts [hunk 1]
```diff
@@ -465,6 +474,14 @@ export default class RoomConnection extends TypedEventTarget {
 
     leave(): Promise<void> {
         return new Promise<void>((resolve) => {
+            if (this._ownsLocalMedia) {
+                this.localMedia?.stop();
+            }
+
+            if (this.rtcManager) {
```
Comment (havardholvik) [inline src/lib/RoomConnection.ts:481]: This is a drive-by fix to properly clean up SFU connections on leave

File: src/lib/RoomConnection.ts [hunk 2]
```diff
@@ -156,6 +159,12 @@ export default class RoomConnection extends TypedEventTarget {
 
         const urls = fromLocation({ host: this.roomUrl.host });
 
+        // Set up local media
```
Comment (sebapawlak) [inline src/lib/RoomConnection.ts:162]: Nit picking here really. My personal preference would be to have a function called `setUpLocalMedia` extracted here rather then the inline comment.

File: src/lib/react/__tests__/useLocalMedia.unit.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import { jest } from "@jest/globals";
```
Comment (sebapawlak) [inline src/lib/react/__tests__/useLocalMedia.unit.ts:1]: is the `.unit` naming convention something new? It's the first time I'm seeing it. Usually we would suffix test files with `.spec`.

File: src/lib/react/__tests__/useLocalMedia.unit.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    beforeEach(() => {
+        MockedLocalMedia.mockClear();
+    });
+
+    it("should assume audio and video by default", () => {
+        renderHook(() => useLocalMedia());
+        expect(MockedLocalMedia.mock.calls[0]).toEqual([{ audio: true, video: true }]);
+    });
+
+    it("should use provided audio and video values", () => {
```
Comment (sebapawlak) [inline src/lib/react/__tests__/useLocalMedia.unit.ts:23]: I would say that we should aim for deterministic unit tests. Can we just test all the cases? These tests are super cheap.
````
    it.each([
        { audio: true, video: true },
        { audio: true, video: false },
        { audio: false, video: true },
        { audio: false, video: false },
    ])(`should use provided audio: $audio and video: $video values`, ({ audio, video }) => {
        renderHook(() => useLocalMedia({ audio, video }));
        expect(MockedLocalMedia.mock.calls[0]).toEqual([{ audio, video }]);
    });
````

**Review by sebapawlak — APPROVED** [review]

File: src/lib/react/__tests__/useLocalMedia.unit.ts (hunk 3 above)
Comment (havardholvik) [reply src/lib/react/__tests__/useLocalMedia.unit.ts:1]: I might just have seen this in some other project of ours, but happy to use spec 👍

File: src/lib/react/__tests__/useLocalMedia.unit.ts (hunk 4 above)
Comment (havardholvik) [reply src/lib/react/__tests__/useLocalMedia.unit.ts:23]: Yea, this was prob an overkill to have the random values. For some tests however, I have found this setup useful in the past.

File: src/lib/RoomConnection.ts (hunk 2 above)
Comment (havardholvik) [reply src/lib/RoomConnection.ts:158]: Will wait with extracting these until we have some more core 😄

File: src/lib/react/__tests__/useLocalMedia.unit.ts (hunk 3 above)
Comment (havardholvik) [reply src/lib/react/__tests__/useLocalMedia.unit.ts:1]: Renamed

File: src/lib/react/__tests__/useLocalMedia.unit.ts (hunk 4 above)
Comment (havardholvik) [reply src/lib/react/__tests__/useLocalMedia.unit.ts]: Cleaned up 👍

File: src/lib/LocalMedia.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+        navigator.mediaDevices.addEventListener("devicechange", this._updateDeviceList.bind(this));
+    }
+
+    addRtcManager(rtcManager: RtcManager) {
+        this._rtcManagers.push(rtcManager);
+    }
+
+    removeRtcManager(rtcManager: RtcManager) {
+        this._rtcManagers = this._rtcManagers.filter((r) => r === rtcManager);
```
Comment (havfo) [inline src/lib/LocalMedia.ts:59]: This is not what you want probably?

```
this._rtcManagers.filter((r) => r !== rtcManager)
```

Comment (havardholvik) [reply src/lib/LocalMedia.ts:59]: Good catch, indeed what we want 😀

File: src/lib/LocalMedia.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        }
+
+        const newValue = enabled ?? !audioTrack.enabled;
+        audioTrack.enabled = newValue;
+
+        this.dispatchEvent(new CustomEvent("microphone_enabled", { detail: { enabled: newValue } }));
+    }
+
+    async setCameraDevice(deviceId: string) {
+        const newStream = await navigator.mediaDevices.getUserMedia({ video: { deviceId } });
```
Comment (havfo) [inline src/lib/LocalMedia.ts:103]: This could throw, is that desired behavior and should be caught by upper layer?

File: src/lib/LocalMedia.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+        this.dispatchEvent(
+            new CustomEvent("stream_updated", {
+                detail: { stream: this.stream },
+            })
+        );
+    }
+
+    private async _updateDeviceList() {
+        const devices = await navigator.mediaDevices.enumerateDevices();
```
Comment (havfo) [inline src/lib/LocalMedia.ts:151]: This could throw

File: src/lib/LocalMedia.ts (hunk 6 above)
Comment (havardholvik) [reply src/lib/LocalMedia.ts:103]: This should be handled here I think, and we should rather emit an error event which the hook could use to update a `state` property to something error-ish.

File: src/lib/LocalMedia.ts (hunk 7 above)
Comment (havardholvik) [reply src/lib/LocalMedia.ts:151]: Same answer as ^ 👍

File: src/lib/LocalMedia.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+                    cameraDevices: devices.filter((d) => d.kind === "videoinput"),
+                    microphoneDevices: devices.filter((d) => d.kind === "audioinput"),
+                    speakerDevices: devices.filter((d) => d.kind === "audiooutput"),
+                },
+            })
+        );
+    }
+
+    async start() {
+        const newStream = await navigator.mediaDevices.getUserMedia(this._constraints);
```
Comment (havfo) [inline src/lib/LocalMedia.ts:165]: This could throw

File: src/lib/RoomConnection.ts [hunk 9]
```diff
@@ -365,23 +374,17 @@ export default class RoomConnection extends TypedEventTarget {
...
-        if (!this.localStream && this.localMediaConstraints) {
-            const localStream = await navigator.mediaDevices.getUserMedia(this.localMediaConstraints);
-            this.localStream = localStream;
-        }
-
-        const organization = await this.organizationServiceCache.fetchOrganization();
-        if (!organization) {
-            throw new Error("Invalid room url");
+        if (!this.localMedia.stream.active) {
+            await this.localMedia.start();
```
Comment (havfo) [inline src/lib/RoomConnection.ts:378]: At the moment, this could throw

File: src/lib/react/useLocalMedia.ts [hunk 10]
```diff
@@ -1,25 +1,119 @@
...
+            };
+        default:
+            return state;
+    }
+}
+
+export default function useLocalMedia(
+    constraints: MediaStreamConstraints = { audio: true, video: true }
+): LocalMediaRef {
+    const [localMedia, setLocalMedia] = useState<LocalMedia | null>(null);
```
Comment (havfo) [inline src/lib/react/useLocalMedia.ts:64]: Why not set the initial `localMedia` here instead of in the useEffect?

File: src/lib/react/useLocalMedia.ts [hunk 11]
```diff
@@ -1,25 +1,119 @@
...
         };
-    }, [localStream]);
+    }, [localMedia]);
 
-    return [localStream];
+    return {
+        state,
+        actions: {
+            setCameraDevice: (...args) => {
+                return localMedia?.setCameraDevice(...args) || Promise.resolve();
```
Comment (havfo) [inline src/lib/react/useLocalMedia.ts:105]: If you set the initial value in `useState`, `localMedia` should be available here always, or am I missing something?

File: src/lib/react/useLocalMedia.ts (hunk 10 above)
Comment (havardholvik) [reply src/lib/react/useLocalMedia.ts:64]: Good question. I tried doing that with the RoomConnection, but that ended up creating new instances multiple times.

I can have another go at cleaning this up 👍

File: src/stories/components/VideoExperience.tsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import useRoomConnection from "../../lib/react/useRoomConnection";
+
+export default function VideoExperience({
+    displayName,
+    roomName,
+    localMedia,
+}: {
+    displayName?: string;
+    roomName: string;
+    localStream?: MediaStream;
```
Comment (havfo) [inline src/stories/components/VideoExperience.tsx:13]: Remnant from some testing?

File: src/lib/react/useLocalMedia.ts (hunk 11 above)
Comment (havardholvik) [reply src/lib/react/useLocalMedia.ts:105]: Yes, it should be. Ill try tweaking this, as it is much cleaner without the optional chaining.

**Comment by havfo** [comment]
I have gone through everything now, see comments. In general looks sane and good :+1: I haven't actually tested it, I will leave that to the other people in the project :slightly_smiling_face:

**Comment by havardholvik** [comment]
> I have gone through everything now, see comments. In general looks sane and good :+1: I haven't actually tested it, I will leave that to the other people in the project :slightly_smiling_face: 

Thanks for the review 🙏 Ill try fixing the `useState` usage you mentioned. 

For the error handling Ill come up with a draft on how we could model the error state. Might be a separate PR that though 👍

File: src/lib/react/useLocalMedia.ts (hunk 10 above)
Comment (havardholvik) [reply src/lib/react/useLocalMedia.ts:64]: Hah! Got it to work. The key was to wrap the `new LocalMedia(...)` call in an anonymous function. Found the tip here https://stackoverflow.com/a/64131447. This makes everything much cleaner I agree, as we can always assume there is an instance 👍

File: src/lib/react/useLocalMedia.ts (hunk 11 above)
Comment (havardholvik) [reply src/lib/react/useLocalMedia.ts]: Fixed

**Comment by sebapawlak** [comment]
> I have gone through everything now, see comments. In general looks sane and good 👍 I haven't actually tested it, I will leave that to the other people in the project 🙂

I have tested it and seems to be working fine!

File: src/lib/LocalMedia.ts (hunk 6 above)
Comment (havardholvik) [reply src/lib/LocalMedia.ts:103]: Added handling

File: src/lib/LocalMedia.ts [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+        this.dispatchEvent(
+            new CustomEvent("stream_updated", {
+                detail: { stream: this.stream },
+            })
+        );
+    }
+
+    async setMicrophoneDevice(deviceId: string) {
+        const newStream = await navigator.mediaDevices.getUserMedia({ audio: { deviceId } });
```
[the remark this answers is not in the document]
Comment (havardholvik) [reply src/lib/LocalMedia.ts:127]: Handled

File: src/lib/LocalMedia.ts (hunk 7 above)
Comment (havardholvik) [reply src/lib/LocalMedia.ts]: Handled

File: src/lib/LocalMedia.ts (hunk 8 above)
Comment (havardholvik) [reply src/lib/LocalMedia.ts:178]: Handled 👍

File: src/lib/RoomConnection.ts (hunk 9 above)
Comment (havardholvik) [reply src/lib/RoomConnection.ts:379]: Will fix this separately, as this PR focuses more on the local media files. Before using localMedia in this class, the gUM could still throw, so this flaw has been here all along
